### PR TITLE
Fix CPython bytecode changes in Python 3.6

### PR DESCRIPTION
**Pull request by rosswhitfield**
From Python 3.6 the Python interpreter now uses a 16-bit wordcode instead of 8-bit bytecode. Also some new function call opcodes have been added. See https://docs.python.org/3.6/whatsnew/3.6.html#cpython-bytecode-changes

I also removed the variable `e` since `iarg` is not defined therefore `e` must never change.

Refs #18358

**To test:**
 - [x] @quantumsteve can you check that this also fixes OSX with python 3.6
 - [x] @martyngigg can you have a look at the code since I'm guessing you're the one most familiar with `functinspect.py`

*Does not need to be in the release notes.*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by quantumsteve** [comment]
LGTM. I built this branch on [OS X with python 3.6](http://builds.mantidproject.org/job/master_clean-osx-10.11-python3/165/console) and verified that this addresses many of the python3 issues found in #18358.

**Review by quantumsteve — APPROVED** [review]
LGTM. I built this branch on [OS X with python 3.6](http://builds.mantidproject.org/job/master_clean-osx-10.11-python3/165/console) and verified that this addresses many of the python3 issues found in #18358.

File: Framework/PythonInterface/mantid/kernel/funcinspect.py [hunk 1]
```diff
@@ -106,18 +106,17 @@ def decompile(code_object):
...
     while i < n:
         i_offset = i
         i_opcode = opcode_from_bytecode(code[i])
         i = i + 1
         if i_opcode >= opcode.HAVE_ARGUMENT:
-            i_argument = opcode_from_bytecode(code[i]) + (opcode_from_bytecode(code[i+1]) << (4*2)) + e
-            i = i + 2
-            if i_opcode == opcode.EXTENDED_ARG:
-                e = iarg << 16
+            if sys.version_info < (3, 6): # from 3.6 bytecode is now 16-bit not 8-bit
```
Comment (martyngigg) [inline Framework/PythonInterface/mantid/kernel/funcinspect.py]: The changes make sense. Is it worth caching the version check at the module level to save re-evaluating it?

Comment (rosswhitfield) [inline Framework/PythonInterface/mantid/kernel/funcinspect.py]: Good point. I will do that.

**Comment by serhiy-storchaka** [comment]
The code doesn't support `EXTENDED_ARG`. Thus it works incorrectly for opcode arguments larger than 65535 in Python < 3.6 and larger than 255 in Python 3.6+. It is unlikely that a function uses more than 65535 constants or has jumps to an offset larger than 65535, but it is easier to exceed the limit of 255.

I would suggest to use `dis.Bytecode()` or `dis.get_instructions()` (available since Python 3.4).

**Comment by martyngigg** [comment]
@serhiy-storchaka Thanks for the comments around lack of support for `EXTENDED_ARG`.

Our set of use cases for this are limited so this is not currently a problem but I think we'll look to improve the code for Python >= 3.4 in the future to use `dis`.

**Review by martyngigg — APPROVED** [review]

**Comment by martyngigg** [comment]
:shipit:

**Pull request merged**